include/internal/hashtable.h

```diff
@@ -0,0 +1,319 @@
+{                                                                              \
+    if (v == NULL)                                                             \
+        return NULL;                                                           \
+                                                                               \
+    *v = ossl_ht_get(h, key);                                                  \
+    if (*v == NULL)                                                            \
+        return NULL;                                                           \
+    return ossl_ht_##name##_##vtype##_from_value(*v);                          \
+}                                                                              \
+                                                                               \
+pfx ossl_unused HT_VALUE *ossl_ht_##name##_##vtype##_to_value(vtype *data,     \
+                                                              HT_VALUE *v)     \
+{                                                                              \
+    v->type_id = name##_##vtype##_id;                                          \
+    v->value = data;                                                           \
+    return v;                                                                  \
+}                                                                              \
+                                                                               \
+pfx ossl_unused int ossl_ht_##name##_##vtype##_type(HT_VALUE *h)               \
+{                                                                              \
+    return h->type_id == name##_##vtype##_id;                                  \
+}
+
+#define DECLARE_HT_VALUE_TYPE_FNS(vtype, name)                                 \
+int ossl_ht_##name##_##vtype##_insert(HT *h, HT_KEY *key, vtype *data,         \
+                                      vtype **olddata);                        \
+vtype *ossl_ht_##name##_##vtype##_from_value(HT_VALUE *v);                     \
+vtype *ossl_unused ossl_ht_##name##_##vtype##_get(HT *h,                       \
+                                                  HT_KEY *key,                 \
+                                                  HT_VALUE **v);               \
+HT_VALUE *ossl_ht_##name##_##vtype##_to_value(vtype *data, HT_VALUE *v);       \
+int ossl_ht_##name##_##vtype##_type(HT_VALUE *h);                              \
+
+/*
+ * Helper function to construct case insensitive keys
+ */
+static inline void ossl_ht_strcase(char *tgt, const char *src, int len)
+{
+    int i;
+#if defined(CHARSET_EBCDIC) && !defined(CHARSET_EBCDIC_TEST)
+    const long int case_adjust = ~0x40;
+#else
+    const long int case_adjust = ~0x20;
+#endif
+
+    if (src == NULL)
+        return;
+
+    for (i = 0; src[i] != '\0' && i < len; i++)
+        tgt[i] = case_adjust & src[i];
+}
+
+/*
+ * Create a new hashtable
+ */
+HT *ossl_ht_new(HT_CONFIG *conf);
```

Comment:
Don't make the config dynamic. LHASH has dynamically specified hash and compare function pointers but I don't like this. In practice it is almost never necessary and this will allow inlining of the hash and compare function calls. Should be part of the generated code for each "instantiation".

Reply:
So, you would prefer that we select a hash function that doesn't provide the caller an opportunity to override it?  Thats fine with me, but note thats not the only config setting (i.e. removing the hash function won't get rid of the config data as a whole).